src/dynsight/_internal/vision/detect.py

```diff
@@ -0,0 +1,782 @@
+from __future__ import annotations
+
+import pathlib
+import random
+import shutil
+import tkinter as tk
+from pathlib import Path
+from typing import TYPE_CHECKING, Literal
+
+import matplotlib.pyplot as plt
+import numpy as np
+import yaml
+from PIL import Image
+from scipy.optimize import curve_fit
+from scipy.stats import norm
+from ultralytics import YOLO
+
+from .vision_gui import VisionGUI
+
+if TYPE_CHECKING:
+    from .video_to_frame import Video
+
+
+class Detect:
+    """A class to manage the full pipeline of object detection from videos.
+
+    Attributes:
+        project_folder: Main directory for all project outputs.
+        frames_dir: Directory to store extracted video frames.
+        training_items_path: Directory to save training item crops selected.
+        syn_dataset_path: Directory for the generated synthetic dataset.
+        models_path: Directory to save trained models.
+        predictions_path: Directory to save prediction outputs.
+        yaml_file: Path to the YAML configuration file for dataset training.
+        video_size: Resolution of the input video (width, height).
+        n_frames: Total number of frames extracted from the video.
+
+    Methods:
+        synthesize(): Generate a synthetic dataset by creating random collages.
+        train(): Train a YOLO model on the synthetic dataset.
+        predict(): Make predictions using a trained model on video frames.
+    """
+
+    def __init__(
+        self,
+        input_video: Video,
+        project_folder: pathlib.Path = Path("output_folder"),
+    ) -> None:
+        """Initialize a detection project.
+
+        Creates all necessary subdirectories and extracts video frames if not
+        already present.
+
+        Args:
+            input_video: The input video object from which to extract frames.
+            project_folder: Path to the main project directory.
+
+        """
+        # Define the project folder path
+        self.project_folder = project_folder
+
+        # Define all the sub folders path
+        self.frames_dir = self.project_folder / "frames"
+        self.training_items_path = self.project_folder / "training_items"
+        self.syn_dataset_path = self.project_folder / "synthetic_dataset"
+        self.models_path = self.project_folder / "models"
+        self.predictions_path = self.project_folder / "predictions"
+
+        # Define the config file path for the training
+        self.yaml_file = self.project_folder / "training_options.yaml"
+
+        # Extract information from the input video
+        self.video_size = input_video.resolution()
```

Comment:
See my comment on `Video`, but yes, loading the video twice to get these values is probably not ideal